modules/platform/nuxeo-platform-web-common/src/test/java/org/nuxeo/ecm/platform/web/requestcontroller/filter/TestNuxeoIdempotentFilter.java

```diff
@@ -0,0 +1,386 @@
+            }).when(response).getHeader(anyString());
+            doAnswer(invocation -> {
+                return headers.get(invocation.getArguments()[0]);
+            }).when(response).getHeaders(anyString());
+            doAnswer(invocation -> {
+                headers.put((String) invocation.getArguments()[0], List.of((String) invocation.getArguments()[1]));
+                return null;
+            }).when(response).setHeader(anyString(), anyString());
+            doAnswer(invocation -> {
+                headers.computeIfAbsent((String) invocation.getArguments()[0], k -> new ArrayList<>())
+                       .add((String) invocation.getArguments()[1]);
+                return null;
+            }).when(response).addHeader(anyString(), anyString());
+        }
+
+        public int getStatus() {
+            return status;
+        }
+
+        public HttpServletResponse getResponse() {
+            return response;
+        }
+
+        public OutputStream getOutput() {
+            return output;
+        }
+
+    }
+
+    @Before
+    public void setUp() throws IOException {
+        filter = new NuxeoIdempotentFilter();
+        chain = mock(FilterChain.class);
+        request = mock(HttpServletRequest.class);
+        mockResponse = new MockResponse();
+        // handle store
+        store = (KeyValueStoreProvider) kvs.getKeyValueStore(NuxeoIdempotentFilter.DEFAULT_STORE);
+        store.clear();
+    }
+
+    @After
+    public void tearDown() {
+        if (filter != null) {
+            filter.destroy();
+        }
+        if (store != null) {
+            store.clear();
+        }
+    }
+
+    protected void checkResponse(MockResponse mockResponse, Integer status, String content, boolean hasHeaders) {
+        assertEquals(status, (Integer) mockResponse.getStatus());
+        assertEquals(content, mockResponse.getOutput().toString());
+        if (hasHeaders) {
+            assertEquals(RESPONSE_HEADERS.keySet(), mockResponse.getResponse().getHeaderNames());
+            RESPONSE_HEADERS.forEach((k, v) -> {
+                assertEquals(v, mockResponse.getResponse().getHeaders(k));
+            });
+        } else {
+            assertTrue(mockResponse.getResponse().getHeaderNames().isEmpty());
+        }
+    }
+
+    protected void checkStore(String status, String content) {
+        assertEquals(content, store.getString(KEY));
+        String ikey = KEY + NuxeoIdempotentFilter.INFO_SUFFIX;
+        if (status == null) {
+            assertNull(store.getString(ikey));
+        } else {
+            if (NuxeoIdempotentFilter.INPROGRESS_MARKER.equals(status)) {
+                assertEquals(status, store.getString(ikey));
+            } else {
+                String info = "{\"headers\":" //
+                        + "{\"Connection\":[\"Keep-Alive\"]," //
+                        + "\"Content-Encoding\":[\"gzip\"]," //
+                        + "\"Content-Type\":[\"text/html; charset=utf-8\"]" //
```

Comment:
Could you add case with a multi-value header? For example `Set-Cookie` and `Accept` are common headers seen multiple times.

Reply:
Fixed the "isFirst" buggy logic with these additional checks :+1: